the-collab-lab/tcl-54-smart-shopping-list
# issue #12 | Sorting items in list by purchase urgency

## Description



### Task:
To have the UI show the user how urgent each item needs to be purchased. Item status can be:
- `Overdue`: You should buy this item soon since it's pass your estimated purchase date and is not yet inactive.
- `Soon`: You've got 7 days or fewer until the *next* estimated purchase date. Purchase it soon!
- `Kind of Soon`: You've got between 7 & 30 days until the *next* estimated purchase date. Purchase it soon-ish!
- `Not Soon`: You've still got 30 days or more until the *next* estimated purchase date. No need to purchase right away.
- `Inactive`: You have not purchased this in 60 days.

## Our current implementation:

#### Note: Much of the following logic explained here have been modified/refactored.

### Client-side:

#### `ListItem.jsx`

**Update:** This has since been updated and the functionality has been moved to `List.jsx`

**Old way:**
-  In the function `getBuyingUrgency`, we create conditionals to reflect the correct tags and temporary color urgency attributes for our items. 
   - For example, the conditional `if (daysUntilNextPurchase) >= 60` means it's been 60 or more days since the estimated next purchase, so the tag "Inactive (has not been purchased recently" will be shown for that item. Same logic goes for the rest of the purchase urgency.
- To get the value of the aforementioned `dayUntilNextPurchase`, the `getDaysBetweenDates` function from `utils` is imported and used to calculate the interval between today and `dateNextPurchased` property.
- We call on `getBuyingUrgency()` on line 83 to assign the value of which will be whatever is the appropriate color and string value associated with the item.

#### `List.jsx`

**Update:** This function (`urgencyData`) has been updated to match the logic of `getBuyingUrgency` which has now been moved from `ListItem` to `List`

**Old way:**
- Have a look at `urgencyData`, in we which filter through our `data` for the names of the items. Now that we have the names, we `sort` through them based on a _**function**_ that defines the sort order. The set of instructions inside this function `comparePurchaseUrgency` tell us the order in which the items should be organized (soon, not soon, etc). 
- `comparePurchaseUrgency` is defined in `firebase.js`.
- On line 91 we no longer map the originally pulled `data` for our list items, and instead use the correctly sorted order of our shopping list from `urgencyData`

### Server-side:

**Update:** Changes have been made to `comparePurchaseUrgency` to account for accurate alphabetization of the items.

#### `firebase.js`

- `comparePurchaseUrgency` function: It's taking two parameters, `item1` and `item2`. In this compare function we are comparing one element (first item) with another element (second item). The comparison determines which one of the two will go first in the list. The function does this for all the names on the list. It essentially "walks" through the list of names we retrieved earlier, and sorts everything according to the conditionals in place. At the end of all this, the expected result is that everything should be ordered based on their urgency.
- Conditionals from `comparePurchaseUrgency`:
```
  if (item1.dateNextPurchased.toDate() < item2.dateNextPurchased.toDate()) {
		    return -1;
	    } else if (
		    item1.dateNextPurchased.toDate() > item2.dateNextPurchased.toDate()
	    ) {
		    return 1;
	    } 
          / *... */
    };
```
We're highlighting some of the code to draw attention to the return value, but please refer to the files for a better look at the entire block of code. 

**Refresher:** Remember that `dateNextPurchased` is a document property in Firestore. We are using a `toDate()` function to convert it to a date that JavaScript can read. Because Firestore dates are...quirky. 🤷🏽‍♀️

   - Note in the code block that there is a `return -1` and a `return 1`. The negative or positive value determines the placement of the items on the list.
      - If the `dateNextPurchased` of `item1` is less than the `dateNextPurchased` of `item2`, then `item1` is placed _**before**_ `item2` on the shopping list (hence `-1`).
      - If the opposite is true -- `item1`'s `dateNextPurchased` is _greater than_ `item2`'s -- then `item1` is placed **_after_** `item2` on the shopping list (hence `1`).
      - This is how the compare function works when we `sort`. See the table in the **"Description"** section of this [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description). 

You may or may not be wondering, where are these return values being used? If you recall, in `List.jsx` earlier, we had the `urgencyData` in which we filter and sort the list. We have now passed the return value of `comparePurchaseUrgency` (`-1` or `1`) to the `sort` method, and our list (`urgencyData`) is now expected to be in order of their urgency.

#### `dates.js`
- We rounded up the returned `daysElapsed` with `Math.ceil` so that the comparison value of the days didn't include decimal values, but a rounded number we can use while sorting comparisons in `firebase.js`

## Todos

~~Figuring out the logic to account for an "overdue" item. (See stretch goals)~~

## Related Issue
Closes #12

## Acceptance Criteria


### Stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
| ✓   | :sparkles: New feature     |


## Updates

### Before

![image](https://user-images.githubusercontent.com/89554728/219322086-11c96cbf-4ebe-4715-88eb-efc5d3424a09.png)

### After

![image](https://user-images.githubusercontent.com/89554728/219322142-1522e5c6-80a9-4893-b73d-3184c62ba1e1.png)

## Testing Steps / QA Criteria

1. From your terminal, first run `git pull` and then `git checkout ac-ad-ui-to-display-item-purchasing-urgency`.
2. Next, run `npm start` to open http://localhost:3000/.
3. Create or join a shopping list.
4. Add a new items to the list and manipulate the dates in [Firebase](https://console.firebase.google.com/u/0/project/tcl-54-smart-shopping-list/firestore/data/~2Fburl%20swirl%20ahem~2FiPIevLRFXvR3vA4PtuY1) to confirm if an item is due to be purchased and is categorized correctly: overdue, soon, kind of soon, not soon, or inactive.
5. Items are alphabetized only if they have the **exact** amount of days until the next purchase date, otherwise they are sorted by soonest to be purchased at the top and latest at the bottom. Overdue items are not alphabetized if they have the exact same amount of days since `dateLastPurchased`.

- Note: The color display and UI of the urgency is only temporary here until we decide as a team what style we'd like to move forward with for categorizing items 🌟
